packages/yew/src/io_coop.rs

```diff
@@ -19,9 +16,17 @@ mod io_tokio {
     where
         F: Future<Output = ()> + 'static,
     {
-        tokio::task::spawn_local(f);
+        #[cfg(feature = "tokio")]
+        ::tokio::task::spawn_local(f);
+        #[cfg(not(feature = "tokio"))]
+        {
+            let _ = f;
+            panic!(
+                r#"No scheduler configured for this platform, features related to async can't be used.
+                Either compile with `target_arch = "wasm32", or enable the `tokio` feature."#
+            );
```

Comment:
`compile_error!` instead of `panic!`?

Reply:
I thought about that: the reason I stuck with `panic!` is that if you don't use `send_future`, or `Suspension::from_future` (i.e. async support), technically you don't need to configure the scheduler. On non-wasm targets, the scheduler starts immediately without scheduling for the next tick.

Reply:
Right. It allows compiling apps that don't use async stuff to be used in non-tokio runtimes

Reply:
I don't think this is the correct way to deal with this issue, this downgrades a previously compile time error to a runtime error.

Also, the scheduler is always available.
What's not available is the runtime to spawn thread local tasks.

Reply:
Suppose I write a library crate targeting ssr. Then that lib should _not_ enable a specific scheduler. Only the final binary should commit to exactly one (currently only tokio, but I suppose we could add other options). I don't see why I'd have to enable a scheduler for it to type-check, even if actually *running* any tests or the final binary requires that.

Maybe the `ServerRenderer` or other entry point can assert one is set, so it fails faster? Also a feature_soundness lint that ensures only one is enabled would make sense in the future.

Reply:
No, `ServerRenderer` does not require any runtime to be present.

Usage of `spawn_local` is not mandatory.

You can even use `futures::executor::block_on` to pull `ServerRenderer` until it completes.

Reply:
I know `ServerRenderer` doesn't require any runtime, which is the point of downgrading this from a compile time to a runtime panic. I thought you wanted the opposite and asserting the presence in one of the ssr entry points would be a trade-off to fail fast, but only when actually running the code as compare to type-checking a lib.

Reply:
> which is the point of downgrading this from a compile time to a runtime panic

The previous code will not look for `spawn_local` at all if no future-based code is enabled.

This change downgrades `send_future` usage in unsupported runtimes (e.g.: `async-std`) from compile time error to runtime error.

Reply:
Have you tried writing a component lib that uses the async features?

```rust
// src/lib.rs
use yew::prelude::*;

#[function_component]
fn UsesAsync() -> Html {
    let _ = yew::suspense::use_future(async {});
    html! {
        <div />
    }
}
// $ cargo check
// error[E0425]: cannot find function `use_future` in module `yew::suspense`
//  --> examples/sample_lib/src/lib.rs:5:28
//   |
// 5 |     let _ = yew::suspense::use_future(async {});
//   |                            ^^^^^^^^^^ not found in `yew::suspense`
```

Oops, I have to install the `wasm32` target and use `--target=wasm32` or enable one of the schedulers to even type check this. That is pointless friction, imo. `wasm_bindgen` does already commit to the solution of providing panicking mocks on platforms that are not supported, and the standard library does too (e.g. for `std::time::Instant` if I recall correctly).

*This* is the reason for just panicking when nothing is configured. It fails with a clear error message and the only party feeling a slight friction is the author of the final *binary*, instead of everybody writing a library in-between.

I'm open to adding more docs, if it should be more clearly explained, but having to take care of feature flags (and perhaps even hiding components behind them in a library, if you really lint for feature soundness as a library author) is just a hassle.

> Why? if it does not even require one.

Good, same opinion here.

Reply:
> Oops, I have to install the wasm32 target and use --target=wasm32 

Isn't this stated as a prerequisite in the documentation?

TBH, if you want to use SSR, it's almost always desirable to use feature flags to prevent any SSR-only code to be included in your final application.

You can hardly write anything useful without runtime dependent-code (even `sleep` is runtime-dependent).

If you are using CSR only, then you should not care about non-wasm targets as your library / application will only run under wasm32.

Reply:
Even if I only care about CSR and wasm, the point is that I have to configure every single tool to be aware of this, if the default target doesn't pretend that the methods are there. The editor has to be aware that I target yew with an async runtime for code completion, the CI has to be setup (even the clippy one, not talking about unit tests here), etc etc. That's friction and since none of this is *meant* to have any alternative semantics anyway, I'd argue that extra friction can be avoided by just providing sensible mocks.

Reply:
OK. If it helps to ease the feature flag mess for library maintainers.